packages/core/contracts/trader/uniswap-broker/UniswapBroker.sol

```diff
@@ -0,0 +1,171 @@
+     * possible to the truePrice. 
+     * @dev True price is expressed in the ratio of token A to token B.
+     * @dev The caller must approve this contract to spend whichever token is intended to be swapped.
+     * @param uniswapRouter address of the uniswap router used to facilate trades.
+     * @param uniswapFactory address of the uniswap factory used to fetch current pair reserves.
+     * @param tokenA address of the first token in the uniswap pair.
+     * @param tokenA address of the second token in the uniswap pair.
+     * @param truePriceTokenA the nominator of the true price.
+     * @param truePriceTokenB the denominatornominator of the true price.
+     * @param maxSpendTokenA maximum to spend in tokenA. Note can be set to zero, thereby limiting the direction of trade.
+     * @param maxSpendTokenA maximum to spend in tokenB. Note can be set to zero, thereby limiting the direction of trade.
+     * @param to recipient of the trade proceeds.
+     * @param to deadline to limit when the trade can execute. If the tx is mined after this timestamp then revert.
+     */
+
+    function swapToPrice(
+        address uniswapRouter,
+        address uniswapFactory,
+        address tokenA,
+        address tokenB,
+        uint256 truePriceTokenA,
+        uint256 truePriceTokenB,
+        uint256 maxSpendTokenA,
+        uint256 maxSpendTokenB,
+        address to,
+        uint256 deadline
+    ) public {
+        IUniswapV2Router01 router = IUniswapV2Router01(uniswapRouter);
+
+        // true price is expressed as a ratio, so both values must be non-zero
+        require(truePriceTokenA != 0 && truePriceTokenB != 0, "SwapToPrice: ZERO_PRICE");
+        // caller can specify 0 for either if they wish to swap in only one direction, but not both
+        require(maxSpendTokenA != 0 || maxSpendTokenB != 0, "SwapToPrice: ZERO_SPEND");
+
+        bool aToB;
+        uint256 amountIn;
+        {
+            (uint256 reserveA, uint256 reserveB) = getReserves(uniswapFactory, tokenA, tokenB);
+            (aToB, amountIn) = computeTradeToMoveMarket(truePriceTokenA, truePriceTokenB, reserveA, reserveB);
+        }
+
+        require(amountIn > 0, "SwapToPrice: ZERO_AMOUNT_IN");
+
+        // spend up to the allowance of the token in
+        uint256 maxSpend = aToB ? maxSpendTokenA : maxSpendTokenB;
+        if (amountIn > maxSpend) {
+            amountIn = maxSpend;
+        }
+
+        address tokenIn = aToB ? tokenA : tokenB;
+        address tokenOut = aToB ? tokenB : tokenA;
+        TransferHelper.safeTransferFrom(tokenIn, msg.sender, address(this), amountIn);
+        TransferHelper.safeApprove(tokenIn, address(router), amountIn);
+
+        address[] memory path = new address[](2);
+        path[0] = tokenIn;
+        path[1] = tokenOut;
+
+        router.swapExactTokensForTokens(
+            amountIn,
+            0, // amountOutMin: we can skip computing this number because the math is tested
```

Comment:
What do you mean by the math is tested?

Reply:
I think this is a comment pulled from the original uniswap implementation.

Reply:
yup. most of this code was copied from the link I referenced in the description.